### ISSUE TITLE
CC: Pull image and unpack of encrypted image on guest

### ISSUE THREAD

**Issue by stevenhorsman** [kata-containers/kata-containers#8111]
This feature builds upon the initial basic image pull work in https://github.com/kata-containers/kata-containers/issues/8103 as one of the possible next steps to extend it to support more of the current `CCv0` feature set.

The scope of this is to add the ability to pull an encrypted container image on the guest and using a provided decryption key decrypt it and unpack it, so we can run an pod from it.

Initially I/we'd hoped to be able to cherry-pick over commits from `CCv0`, but given how much iteration there has been we're going to need to copy over chunks of the code and tests, improving as we go. Ideally we'll add `co-authored` tags to reflect the original authors though.

## Things to note/include with this work
- With this feature (unless already added by something like signature validation if that is picked up before this) we'll need to make the attestation-agent available to image-rs, but I'm not sure we need the confidential-data-hub, or api-server-rest at this point. This involves the follow code changes (plus others I may have missed):
  -  Add attestation-agent to the [versions.yaml](https://github.com/kata-containers/kata-containers/blob/1ad87faaf4d3068569d99c270df640e41465c410/versions.yaml#L200-L203)
  - `rootfs.sh` changes to add the `AA_KBC` parameter (assuming we want to keep this?) and then to [install the attestation-agent](https://github.com/kata-containers/kata-containers/blob/1ad87faaf4d3068569d99c270df640e41465c410/tools/osbuilder/rootfs-builder/rootfs.sh#L683-L697) into the rootfs. See questions
  - code changes in `main.rs` related to [initialising the attestation-agent](https://github.com/kata-containers/kata-containers/blob/1ad87faaf4d3068569d99c270df640e41465c410/src/agent/src/main.rs#L389). This should be toggled on the same feature we add for image_pull support in #8103
  - Adding support for `aa_kbc_params` in agent config: https://github.com/kata-containers/kata-containers/pull/5533/commits/112a3d2baef3709a50655b23d6bd80ef121d50c0
  - The keyprovider pieces of [`get_security_config` in image_rpc.rs](https://github.com/kata-containers/kata-containers/blob/1ad87faaf4d3068569d99c270df640e41465c410/src/agent/src/image_rpc.rs#L143-L152)
- New tests
  - There are three new test cases I'd suggest, that are detailed in the acceptance criteria. The code for this code be based on the [existing `CCv0` tests](https://github.com/kata-containers/tests/blob/CCv0/integration/kubernetes/confidential/agent_image_encrypted.bats), but we want to work to make them easier to apply to different KBCs/TEE by extracting out the KBC/TEE logic into library functions, so the test cases are cleaner, similar to the approach done in https://github.com/kata-containers/kata-containers/blob/main/tests/integration/kubernetes/k8s-confidential.bats rather than a pure cherry-pick
  - For the offline_fs_kbc scenario which is used in non-TEE hardware and SE, we need to have a way to modify the guest image to add the `/etc/aa-offline_fs_kbc-keys.json` file that contains the decryption key. This is [currently done](https://github.com/kata-containers/tests/blob/6cf430642335c010014b5d9355a861328c4213ed/integration/kubernetes/confidential/lib.sh#L129) using the [`cp_to_guest_img`](https://github.com/kata-containers/tests/blob/6cf430642335c010014b5d9355a861328c4213ed/lib/common.bash#L295), which ideally we said we'd prefer not to move across into `main`, but rather use policies to pass through this sort of configuration, but I don't think that is available, so I guess we just need to temporarily port over the cp_to_guest_img and associated lib functions over, if it hasn't already been moved over as part of another issue (like #8101)

## Questions
- How do we control and create the attestation-agent integration in the rootfs? If it is small, should we always include it, or do we need separate versions of the image with AA installed? How about support for mutliple KBCs? Currently I don't think there are any plans to support a single attestation-agent binary with multiple KBC modules built, so does this mean that we need different images build for each KBC we support, which will make kata-deploy larger and potential more painful with extra runtime classes? @Xynnn007 @fidencio 
- Does decryption need the CDH & api-server-rest? or are they still just needed for sealed-secrets? @arronwy 
- In `CCv0` we only have integration tests for this feature on Kubernetes, so do we want and can we test this on `crictl` as well?
- What container image do we want to use? In `CCv0` we use, `[docker.io/katadocker/ccv0-ssh](https://hub.docker.com/r/katadocker/ccv0-ssh)` built by https://github.com/confidential-containers/confidential-containers/tree/main/demos/ssh-demo, but should we take ownership of the image and build it ourselves like we do with the unencrypted test image: https://github.com/kata-containers/kata-containers/tree/main/tests/integration/kubernetes/runtimeclass_workloads/confidential? 

## Acceptance Criteria

For each supported platform (different TEEs):

_**Scenario: Pull encrypted image on guest works**_
**Given** I have a version of kata deployed with a guest image that has an agent with `guest_pull` feature enabled and nydus-snapshotter installed and configured for [guest-pulling](https://github.com/containerd/nydus-snapshotter/blob/main/misc/snapshotter/config-coco-guest-pulling.toml)
  **And** A public encrypted container image *i* with a decryption key *k* that is configured in a KBS, so that image-rs on the guest can connect to it
**When** I try and create a pod from *i*
**Then** The pod is successfully created and runs

_**Scenario: Cannot pull encrypted image with no decryption key**_
**Given** I have a version of kata deployed with a guest image that has an agent with `guest_pull` feature enabled and nydus-snapshotter installed and configured for [guest-pulling](https://github.com/containerd/nydus-snapshotter/blob/main/misc/snapshotter/config-coco-guest-pulling.toml)
  **And** A public encrypted container image *i* with a decryption key *k* that is **not** configured in a KBS, so that image-rs on the guest can connect to it
**When** I try and create a pod from *i*
**Then** The pod is not created an errors (with a helpful message hopefully)

_**Scenario: Cannot pull encrypted image with wrong decryption key**_
**Given** I have a version of kata deployed with a guest image that has an agent with `guest_pull` feature enabled and nydus-snapshotter installed and configured for [guest-pulling](https://github.com/containerd/nydus-snapshotter/blob/main/misc/snapshotter/config-coco-guest-pulling.toml)
  **And** A public encrypted container image *i* with a decryption key *k* and a different key *k'* that is configured in a KBS, so that image-rs on the guest can connect to it
**When** I try and create a pod from *i*
**Then** The pod is not created an errors (with a helpful message hopefully)

**Comment by Xynnn007** [comment]
> How do we control and create the attestation-agent integration in the rootfs?

Before replying to the question, one thing to mention is that we will do some refactoring upon AA and CDH. We will move the image decryption and signature verication abilities from AA to CDH. Thus we need to think about the two different components. Currently we have combinations (`offline-fs-kbc` is by default supported):
- `offline-fs-kbc` CDH without AA
- `cc_kbc+offline-fs-kbc` CDH + AA (TDX/SNP/AZ VTPM SNP/CCA)
- `online-sev-kbc+offline-fs-kbc` CDH without AA

For the question, my idea is to have a flag named `TEE_PLATFORM` to replace `AA_KBC`. The reason is that we do not need to specify KBCes anymore. We had some different KBCs before, but some of these are abandoned:
- `Sample KBC`: Abandoned
- `Offline-fs-kbc`: Used in tests and non-tee environments
- `eaa-kbc`: Abandoned
- `Online-sev-kbc`: Only used in SE
- `offline-sev-kbc`: Abandoned
- `cc-kbc`: used in SGX/TDX/SNP/AZ-VTPM-SNP/CCA

we can easily find that if the platform is decided, the kbc type is decided
- TDX: cc_kbc_tdx
- SNP: cc_kbc_snp
- test/non-tee: offline-fs-kbc
- SGX: cc_kbc_sgx
- Azure vTPM on SNP: cc_kbc_az_vtpm_snp
- CCA: cc_kbc_cca
- ES: online-sev-kbc

Different platforms would also need to install different dependencies, like [tdx](https://github.com/kata-containers/kata-containers/blob/1ad87faaf4d3068569d99c270df640e41465c410/tools/osbuilder/rootfs-builder/ubuntu/config.sh#L41-L56)

With the flag `TEE_PLATFORM` when building image, we can both install dependencies and CDH+AA combination, e.g.
`TEE_PLATFORM=TEST`: `offline-fs-kbc` CDH without AA
`TEE_PLATFORM=tdx`: `cc-kbc+offline-fs-kbc` CDH + AA (TDX)
`TEE_PLATFORM=snp`: `cc-kbc+offline-fs-kbc` CDH + AA (SNP)
...
`TEE_PLATFORM=es`: `online-sev-kbc+offline-fs-kbc` CDH without AA

> If it is small, should we always include it, or do we need separate versions of the image with AA installed?

As mentioned, I suggest we have separate versions of the image.

> How about support for mutliple KBCs?

A universal image to support multiple platform can help a user to quick deploy CoCo without considering about the underlying TEE type, a real user might never use a such one as it will have a bigger image with all different platform dependencies installed. My idea is we can consider two different scenarios
- For CI tests: we use `offline-fs-kbc`, s.t. `TEE_PLATFORM=TEST` image to test common functionalities like image unpack
- For release: we release different images for different platforms. Furthermore, we can publish reference values of the measurements together with the images.

> Currently I don't think there are any plans to support a single attestation-agent binary with multiple KBC modules built, so does this mean that we need different images build for each KBC we support, which will make kata-deploy larger and potential more painful with extra runtime classes?

I am not an expert upon this. I'd like to listen to others' opinions : )

**Comment by stevenhorsman** [comment]
> For the question, my idea is to have a flag named TEE_PLATFORM to replace AA_KBC. The reason is that we do not need to specify KBCes anymore. We had some different KBCs before, but some of these are abandoned:

I like this idea. One extra addition is that offline_fs_kbc is used in SecureExecution as well as the non-TEE, so I guess we want to have a `TEE_PLATFORM` for that too?

> As mentioned, I suggest we have separate versions of the image.

I think this is potentially more viable if we run the AA builds aligning with the different architectures. One of my big concerns was having 6 KBC options * n TEE/architecture options, blowing the problem up a lot, but if we have a single AA type per TEE then I don't think that's unmanageable (in kata-containers main CI we already build 5/6 images I think) and with something like NFD could make the user experience less painful too. @fidencio - Are you happy with this approach and then we can update the description of the task?

**Comment by Xynnn007** [comment]
> I like this idea. One extra addition is that offline_fs_kbc is used in SecureExecution as well as the non-TEE, so I guess we want to have a `TEE_PLATFORM` for that too?

By default `offline-fs-kbc` is supported by CDH, and the real behavior is determined by `aa_kbc_params`. Let me fix the reply I made.



> One of my big concerns was having 6 KBC options * n TEE/architecture options

We will only have `n TEE/architecture` options rather than `6 KBC * n TEE/architecture` as the KBC of a specific TEE is determined before build.

**Comment by stevenhorsman** [comment]
Ding, is there an issue on guest-components for this `TEE_PLATFORM` change (and deprecation/removal of the `KBC` parameter in the Makefile) as I guess that any work to integrate the guest-components (AA, CDH, ASR) into rootfs on `main` should wait until that is completed?

**Comment by Xynnn007** [comment]
> Ding, is there an issue on guest-components for this TEE_PLATFORM change (and deprecation/removal of the KBC parameter in the Makefile) as I guess that any work to integrate the guest-components (AA, CDH, ASR) into rootfs on main should wait until that is completed?

Agreed. Let me put an issue and call for comments.

**Comment by Xynnn007** [comment]
https://github.com/confidential-containers/guest-components/issues/373

**Comment by fitzthum** [comment]
I think this makes sense in general, though. At one point we had thought that we could build an AA/CDH that would include support for any platform, but this doesn't seem that important these days given that we are building different rootfs for different platforms anyway.

If we are going to configure CDH/AA at the platform level, I wonder if we should do the same for ASR. I think currently it has a configuration file that is read at runtime, but it might be simpler if we just made this into a build time configuration that was triggered by the different TEE platforms? I guess what we have now works fine, but I feel like the configuration space is a bit complicated.

**Comment by Xynnn007** [comment]
@stevenhorsman Hi Steve, is there anyone working on porting attestation-agent to main? cc @LindaYu17

**Comment by stevenhorsman** [comment]
> @stevenhorsman Hi Steve, is there anyone working on porting attestation-agent to main? cc @LindaYu17

Not to my knowledge - most of the focus is currently on guest pull, so we could look at attestation-agent as part of this issue which would be blocked by basic guest pull, or we could start working on it straight away as part of sealed secrets support, which I don't think has any dependency on the guest pull features

**Comment by ariel-adam** [comment]
@stevenhorsman to who should this issue be assigned to?

**Comment by stevenhorsman** [comment]
> @stevenhorsman to who should this issue be assigned to?

This is blocked, so doesn't make sense to assign to anyone

**Comment by ariel-adam** [comment]
Makes sense.
I've added a new status called "blocked" and tagged this issue as blocked. 
Please change any other blocked issues to this tag as well

**Comment by fidencio** [comment]
@Xynnn007, one thing that's not totally clear for me is the **why** of having those decided at build time.

Let's consider, for intsance, it running in a distro, instead of inside the Kata Containers rootfs.  Ideally what the distro would like to ship is one binary that, at runtime, would be able to decide which KBC it'd use.

IMHO, the fact we need different rootfses now is really not a reasonable justification for us to make it even harder to reduce the amount of rootfses in the future.  Also, I think that folks interested on using this outsiude of the Kata Containers scope would have an easier time packaging everything if it's not done in this very specific way.

**Comment by Xynnn007** [comment]
@fidencio I think the core issue is KBCs. Let me try to explain.

Currently we only have four KBCs in use (eaa kbc has been  **depreciated** in kata-cc and enclave-cc), they are:
- `Sample KBC`: This is **depreciated** in Kata-cc. It is still used in enclave-cc, where we do not need a guest image.
- `Offline Fs Kbc`: This is widely used in CI and s390x scenarios, etc. This is **always built-in** [in CDH](https://github.com/confidential-containers/guest-components/blob/main/confidential-data-hub/kms/src/plugins/kbs/mod.rs#L43) now.
- `Online Sev KBC`: This is only for sev-es where pre-attestation is used. 
- `Cc KBC`: This is used on `TDX/SNP/CCA/Azure series/..`. It is now **arch independent** due to the [finished refactoring](https://github.com/confidential-containers/guest-components/issues/412). This means the code is always the same.

So the left thing is `Online Sev KBC` and `CC KBC`. Current rootfs building, taking TDX for example, we use an env when building the rootfs to [specify the kbc type](https://github.com/kata-containers/kata-containers/blob/CCv0/tools/osbuilder/rootfs-builder/rootfs.sh#L764-L765) like `cc-kbc-tdx`. The reason is in old times, we have different KBCs to specify. Now we could treat `cc-kbc-tdx` as `cc-kbc` in CDH and `tdx-attester` in AA. In CDH side, we only have `online-sev-kbc` vs `cc-kbc` to choose. What I suggested in this issue is that we could be able to care less about what kbc is used -- as mentioned, the currently remaining kbc options indicate that whatever the KBC type could be determined by the arch. Instead, we could use a very simple flag [TEE_PLATFORM](https://github.com/confidential-containers/guest-components/blob/main/Makefile#L21) to make underlying things transparent. This meas
- if `TEE_PLATFORM=sev`, use CDH (online-sev-kbc + offline-fs-kbc)
- if `TEE_PLATFORM=tdx/snp/cca/azure-*/..`, use CDH (cc-kbc + offline-fs-kbc) + AA ($TEE_PLATFORM-attester)

About your viewpoint that "Build a multi-* supported, selected at runtime" principle, I totally agree. The only thing is a trade-off:

If we want to make a unified image to support sev and other architectures, we must install packages corresponding to various architectures in rootfs, such as [tdx's libtdx-attest-dev](https://github.com/kata-containers/kata-containers/blob/CCv0/tools/osbuilder/rootfs-builder/ubuntu/rootfs_lib.sh#L68). This can be a selection, the only shortage is the rootfs size. But for generality purposes this should be fine. That means

Unified image = CDH (cc-kbc + online-sev-kbc) + AA (with all attesters enable) + dependent packages for all TEEs

I am not sure if this makes sense?

**Comment by fitzthum** [comment]
@Xynnn007 why don't we add a an option to the guest components Makefile to build have `TEE_PLATFORM=all`. Also, we should have an option to build SEV and SNP since they share a rootfs.

As you say it is a tradeoff in terms of memory size. Let's just leave that to kata maintainers to decide if they want a different (smaller) AA for each platform or one bigger one.

**Comment by Xynnn007** [comment]
Added the support for `TEE_PLATFORM=all` and `TEE_PLATFORM=amd` in https://github.com/confidential-containers/guest-components/pull/453

**Comment by wainersm** [comment]
Hi @stevenhorsman @Xynnn007 @ChengyuZhu6 @fidencio @fitzthum @GabyCT @beraldoleal (please help on tagging whoever may be interested on this topic):

Given that PR #8484 is merged and #8870 is functionally working; I'm willing to start looking at enabling guest pull decryption. I'm most interested on the CI/testing pieces of this item actually. But I'd like to ensure that I'm not stepping on the anyone's foot. So, is this item being worked by someone already? If yes then what can I do to help? Else let's build a temporary task-force to address this?

**Comment by stevenhorsman** [comment]
I'm not personally aware of anyone working on it, so thanks for stepping up to take a look/lead!

**Comment by wainersm** [comment]
<snip>

>     * What container image do we want to use? In `CCv0` we use, `[docker.io/katadocker/ccv0-ssh](https://hub.docker.com/r/katadocker/ccv0-ssh)` built by https://github.com/confidential-containers/confidential-containers/tree/main/demos/ssh-demo, but should we take ownership of the image and build it ourselves like we do with the unencrypted test image: https://github.com/kata-containers/kata-containers/tree/main/tests/integration/kubernetes/runtimeclass_workloads/confidential? One thing worth noting with this is that to avoid sha caching/matching, we want to ensure that the content of our encrypted image is slightly different to the unencrypted one. We also want to ensure it's a multi-arch manifest container image so that we can easily work in multiple platforms.
> 

I'm looking at  how to generate the encrypted image for the tests same way it is currently done for the unencrypted. 

@stevenhorsman I need to understand this: "One thing worth noting with this is that to avoid sha caching/matching, we want to ensure that the content of our encrypted image is slightly different to the unencrypted one" as initially I thought in simply encrypt the [unencrypted test image](https://github.com/kata-containers/kata-containers/tree/main/tests/integration/kubernetes/runtimeclass_workloads/confidential). Are you referring to the bug on containerd that some layers are still cached on local host then messing with pull on guest?

I've other questions regarding the test image so I should open an issue to track that work and keep the discussion separated.

**Comment by fitzthum** [comment]
A little bit more detail on the caching thing: 

Because of the way containerd caching works, if you pull an unencrypted container image on the host and then pull an encrypted version of the same image on the host, both images will run, even if you don't have access to the keys for the encrypted image. This is because containerd has a per-layer image cache. Encrypted container images contain the hash of the plaintext of each layer in the manifest somewhere (slightly sketchy). When you pull an encrypted image on the host, it will check the layer cache and if an identical layer has already been downloaded, it will be used, and decryption/attestation steps will be skipped. 

Of course we don't pull images on the host, so this isn't a direct concern, but it is still a somewhat nefarious corner case. Imagine that our CI breaks in a way that we accidentally fallback to pulling on the host. We have some checks in the tests to detect this (and in the future we might disable the shared_fs entirely), but if they fail it would seem like one absolute guarantee is that host pulling of an encrypted image should never work. The host should never have the image decryption key so there should be absolutely no way that an encrypted image could run if pulled on the host. Because of caching, there is, if the encrypted image and unencrypted image have the same plaintext layers.

So this isn't a showstopper and maybe it's not worth the confusion of having two different images. It is only significant in cases where multiple things break, but I have been completely bamboozled by this in the past. In general I think it's also generally good practice to not encrypt totally generic images.

**Comment by Xynnn007** [comment]
Hi, I'd like to take this. There is one question that since #8484 is merged and #8120 and #8122 are in front, there are two ways to implement this issue and also the other two and I'd like to get opinions of the way to go.
1. Use original way of CCv0, s.t. add extra parameters of kata-agent from kernel cmdline to control the image pulling features (registry auth, policy, etc.)
2. As proposed in #9266, let's move image-rs to CDH and all image-pulling parameters to CDH from kata-agent. Also use a workaround to parse them from kernel cmdline. Once initdata is prepared, deprecate the workaround.

Personally I prefer way 1 for now. Once a well-defined release with #8120 and #8122 implemented is released, we'd better think about way 2 and start to do some refactoring. The reason is that we should first make the things work as previous, and then make it better.

Any suggestions?

**Comment by wainersm** [comment]
Hi @Xynnn007 ! Glad to see this item will be in very good hands! :) Will you need me to code the CI/testing parts? If yes then I will start breaking it in smaller issues.

**Comment by Xynnn007** [comment]
> Hi @Xynnn007 ! Glad to see this item will be in very good hands! :) Will you need me to code the CI/testing parts? If yes then I will start breaking it in smaller issues.

Hey @wainersm , it would be great if I could get your help!

**Comment by fitzthum** [comment]
@Xynnn007 I agree with starting from the CCv0 approach. Once we have all that in we can refactor and pick up the init-data stuff.

**Comment by wainersm** [comment]
Hi @fitzthum @Xynnn007 ,

> @Xynnn007 I agree with starting from the CCv0 approach. Once we have all that in we can refactor and pick up the init-data stuff.

In general I'm okay with that approach of getting a functional version merged then do incremental updates/improvements. In this case, however, I'm worried about when we switch out of CCv0 approach (i.e. passing via kernel params) to the new one, it will be a break. We will need to either absorb the breakage (document...etc) or add backward-compatibility (so increasing maintenance and testing).

**Comment by stevenhorsman** [comment]
@Xynnn007 - like with #9904 if tests fail with are missing stacktrace:
```
Stack backtrace:
#    0: <unknown>
#    1: <unknown>
#    2: <unknown>
#    3: <unknown>
#    4: <unknown>
#    5: <unknown>
#    6: <unknown>
#    7: <unknown>
#    8: <unknown>
#    9: <unknown>
#   10: <unknown>
#   11: <unknown>
#   12: <unknown>
#   13: <unknown>
#   14: <unknown>
#   15: <unknown>
#   16: <unknown>
```
which must be related to the image-rs/anyhow bump. Any ideas?

**Comment by Xynnn007** [comment]
@stevenhorsman Are these logs from kata-agent's unit tests? The `anyhow` upgration is for builtable for image, and [commit](https://github.com/kata-containers/kata-containers/pull/9904/commits/4cc672a2eb5a41c5c11f9a68bdf7aa96bc76456a) only influences the unit tests of kata-agent. 

If for e2e test, maybe we can try @wainersm 's comment https://github.com/kata-containers/kata-containers/pull/9904#issuecomment-2221414256 and see if anything wrong with resource retrievement.

**Comment by stevenhorsman** [comment]
> @stevenhorsman Are these logs from kata-agent's unit tests? The `anyhow` upgration is for builtable for image, and [commit](https://github.com/kata-containers/kata-containers/pull/9904/commits/4cc672a2eb5a41c5c11f9a68bdf7aa96bc76456a) only influences the unit tests of kata-agent.
> 
> If for e2e test, maybe we can try @wainersm 's comment [#9904 (comment)](https://github.com/kata-containers/kata-containers/pull/9904#issuecomment-2221414256) and see if anything wrong with resource retrievement.

Sorry, I posted this in the wrong place, so moved the conversation to #9382
